csharp/src/Google.Protobuf/Collections/RepeatedField.cs

```diff
@@ -320,10 +341,20 @@ public void Add(IEnumerable<T> values)
             {
                 throw new ArgumentNullException("values");
             }
-            // TODO: Check for ICollection and get the Count, to optimize?
-            foreach (T item in values)
+
+            ICollection collection = values as ICollection;
+            if (collection != null)
             {
-                Add(item);
+                this.EnsureSize(this.count + collection.Count);
+                collection.CopyTo(this.array, this.count);
```

Comment:
I agree consistency would be good - although `typeof(T).IsValueType` isn't the same thing, due to nullable types. Will look at refactoring that at the same time.

I *may* be able to get to this today, but it may well be Monday.